modules/system/classes/asset/BundleManager.php

```diff
@@ -0,0 +1,253 @@
+     */
+    public function loadRegisteredBundles(): void
+    {
+        foreach ($this->callbacks as $callback) {
+            $callback($this);
+        }
+    }
+
+    /**
+     * Returns a list of the registered asset bundles.
+     */
+    public function listRegisteredBundles(): array
+    {
+        if (empty($this->registeredBundles)) {
+            $this->loadRegisteredBundles();
+        }
+
+        return $this->registeredBundles;
+    }
+
+    /**
+     * Get all bundles configured
+     */
+    public function getBundles(): array
+    {
+        return array_keys($this->listRegisteredBundles());
+    }
+
+    /**
+     * Get the packages for a bundle, with compiler specific packages
+     */
+    public function getBundlePackages(string $name, string $assetType): array
+    {
+        $config = $this->listRegisteredBundles()[$name] ?? [];
+
+        $packages = [];
+        foreach ($config as $key => $value) {
+            // Skip handlers
+            if (in_array($key, [static::HANDLER_SETUP, static::HANDLER_SCAFFOLD])) {
+                continue;
+            }
+
+            // Merge in any compiler specific packages for the current compiler
+            if (is_array($value)) {
+                if ($key === $assetType) {
+                    $packages = array_merge($packages, $value);
+                }
+                continue;
+            }
+
+            $packages[$key] = $value;
+        }
+
+        return $packages;
+    }
+
+    /**
+     * Registers a callback function that defines asset bundles. The callback function
+     * should register bundles by calling the manager's registerBundles() function.
+     * This instance is passed to the callback function as an argument. Usage:
+     *
+     *     BundleManager::registerCallback(function ($manager) {
+     *         $manager->registerAssetBundles([...]);
+     *     });
+     *
+     */
+    public function registerCallback(callable $callback): void
+    {
+        $this->callbacks[] = $callback;
+    }
+
+    /**
+     * Registers asset bundles.
+     */
+    public function registerBundles(array $definitions)
+    {
+        $this->registeredBundles = array_replace_recursive($this->registeredBundles, $definitions);
+    }
+
+    /**
+     * Registers a single asset bundle.
+     */
+    public function registerBundle(string $name, array $definition): void
+    {
+        $this->registerBundles([$name => $definition]);
+    }
+
+    /**
+     * Registers a single bundle setup handler.
+     */
+    public function registerSetupHandler(string $name, Closure $closure): void
+    {
+        $this->registeredBundles[$name][static::HANDLER_SETUP] = $closure;
+    }
+
+    /**
+     * Registers a single bundle scaffold handler.
+     */
+    public function registerScaffoldHandler(string $name, Closure $closure): void
+    {
+        $this->registeredBundles[$name][static::HANDLER_SCAFFOLD] = $closure;
+    }
```

Comment:
Same as above, maybe some documentation of the API of handlers would be useful here.